crates/rpc/rpc/src/eth/api/transactions.rs

```diff
@@ -865,6 +881,108 @@ where
         };
         build_transaction_receipt_with_block_receipts(tx, meta, receipt, &all_receipts)
     }
+
+    /// Helper function for `eth_getTransactionReceipt` (optimism)
+    ///
+    /// Returns the receipt
+    #[cfg(feature = "optimism")]
+    pub(crate) async fn build_transaction_receipt(
+        &self,
+        tx: TransactionSigned,
+        meta: TransactionMeta,
+        receipt: Receipt,
+    ) -> EthResult<TransactionReceipt> {
+        // get all receipts for the block
+        let all_receipts = match self.cache().get_receipts(meta.block_hash).await? {
+            Some(recpts) => recpts,
+            None => return Err(EthApiError::UnknownBlockNumber),
+        };
+        let block = self
+            .provider()
+            .block_by_number(meta.block_number)?
+            .ok_or(EthApiError::UnknownBlockNumber)?;
+
+        let l1_block_info = reth_revm::optimism::extract_l1_info(&block).ok().map(Rc::new);
+        let optimism_tx_meta = self.build_op_tx_meta(&tx, l1_block_info, block.timestamp)?;
+
+        build_transaction_receipt_with_block_receipts(
+            tx,
+            meta,
+            receipt,
+            &all_receipts,
+            optimism_tx_meta,
+        )
+    }
+
+    #[cfg(feature = "optimism")]
+    pub(crate) fn build_op_tx_meta(
+        &self,
+        tx: &TransactionSigned,
+        l1_block_info: Option<Rc<L1BlockInfo>>,
+        block_timestamp: u64,
+    ) -> EthResult<OptimismTxMeta> {
+        if let Some(l1_block_info) = l1_block_info {
+            let envelope_buf: Bytes = {
+                let mut envelope_buf = bytes::BytesMut::default();
+                tx.encode_enveloped(&mut envelope_buf);
+                envelope_buf.freeze().into()
+            };
+
+            let l1_fee = (!tx.is_deposit())
+                .then(|| {
+                    l1_block_info.l1_tx_data_fee(
+                        self.inner.provider.chain_spec(),
+                        block_timestamp,
+                        &envelope_buf,
+                        tx.is_deposit(),
+                    )
+                })
+                .transpose()
+                .map_err(|_| EthApiError::InternalEthError)?;
+            let l1_data_gas = (!tx.is_deposit())
+                .then(|| {
+                    l1_block_info.l1_data_gas(
+                        self.inner.provider.chain_spec(),
+                        block_timestamp,
+                        &envelope_buf,
+                    )
+                })
+                .transpose()
+                .map_err(|_| EthApiError::InternalEthError)?;
```

Comment:
these can be merged into a single `then`

Reply:
Merging this into a single then isn't very nice since the `transpose` requires the inner `Option<T>` type `T` to be of result and then we want to distribute the optionality into the inner tuple elements.

I went ahead and merged them in the most idiomatic way I could. Hoping we could not make this a blocker 🙏

Reply:
This is addressed in https://github.com/anton-rs/op-reth/pull/155